core/rs/core/src/tableinfo.rs

```diff
@@ -0,0 +1,261 @@
+ * TableInfo is a struct that represents the results
+ * of pragma_table_info, pragma_index_list, pragma_index_info on a given table
+ * and its indices as well as some extra fields to facilitate crr creation.
+ */
+pub fn pull_table_info(
+    db: *mut sqlite::sqlite3,
+    table: &str,
+    table_info: *mut *mut crsql_TableInfo,
+    err: *mut *mut c_char,
+) -> Result<ResultCode, ResultCode> {
+    let sql = format!("SELECT count(*) FROM pragma_table_info('{table}')");
+    let columns_len = match db.prepare_v2(&sql).and_then(|stmt| {
+        stmt.step()?;
+        stmt.column_int(0)?.to_usize().ok_or(ResultCode::ERROR)
+    }) {
+        Ok(count) => count,
+        Err(code) => {
+            err.set(&format!("Failed to find columns for crr -- {table}"));
+            return Err(code);
+        }
+    };
+
+    let sql = format!(
+        "SELECT \"cid\", \"name\", \"type\", \"notnull\", \"pk\"
+         FROM pragma_table_info('{table}') ORDER BY cid ASC"
+    );
+    let column_infos = match db.prepare_v2(&sql) {
+        Ok(stmt) => {
+            let mut cols: Vec<crsql_ColumnInfo> = vec![];
+
+            while stmt.step()? == ResultCode::ROW {
+                cols.push(crsql_ColumnInfo {
+                    type_: stmt.column_text(2).map_err(free_cols(&cols))?.into_c_ptr(),
+                    name: stmt.column_text(1).map_err(free_cols(&cols))?.into_c_ptr(),
+                    notnull: stmt.column_int(3)?,
+                    cid: stmt.column_int(0)?,
+                    pk: stmt.column_int(4)?,
+                });
+            }
+
+            if cols.len() != columns_len {
+                err.set(&"Number of fetched columns did not match expected number of columns");
+                return Err(free_cols(&cols)(ResultCode::ERROR));
+            }
+            cols
+        }
+        Err(code) => {
+            err.set(&format!("Failed to prepare select for crr -- {table}"));
+            return Err(code);
+        }
+    };
+
+    let (mut pks, non_pks): (Vec<_>, Vec<_>) =
+        column_infos.clone().into_iter().partition(|x| x.pk > 0);
+    pks.sort_by_key(|x| x.pk);
+
+    unsafe {
+        *table_info = crsql_TableInfo {
+            baseCols: column_infos.into_c_ptr(),
+            baseColsLen: columns_len as i32,
+            tblName: table.into_c_ptr(),
+            nonPksLen: non_pks.len() as i32,
+            nonPks: non_pks.into_c_ptr(),
+            pksLen: pks.len() as i32,
+            pks: pks.into_c_ptr(),
+        }
+        .into_c_ptr();
+    }
+
+    return Ok(ResultCode::OK);
+}
+
+fn free_cols<'a>(cols: &'a Vec<crsql_ColumnInfo>) -> impl Fn(ResultCode) -> ResultCode + 'a {
+    move |err: ResultCode| {
+        for info in cols {
+            drop(unsafe { CString::from_raw(info.type_) });
+            drop(unsafe { CString::from_raw(info.name) });
+        }
+        err
+    }
+}
+
+pub unsafe fn free_table_info(table_info: *mut crsql_TableInfo) {
```

Comment:
I think we're pretty close now to being able to move `crsql_TableInfo` into Rust and then we can start removing these explicit frees.

Reply:
That would make things so much nicer